otopimdp/parser.py

```diff
@@ -221,6 +256,56 @@ def _process_event(self, event_type, attributes):
                     break
             attributes['value'] = lines
 
+        if event_type == QUERY_FRAME_EVENT:
+            self._process_frame_event(event)
+
+    def _process_frame_event(self, event):
+        attributes = event[ATTRIBUTES_KEY]
+        while True:
+            line = self.next_line()
+            m = QUERY_FRAME_PATTERNS[QUERY_FRAME_PART_END].match(line)
+            if m:
+                if m.group('frame_name') != attributes['frame_name']:
+                    self.logger.warning(
+                        "The QStart:NAME != QEnd:NAME (%s != %s)",
+                        m.group('frame_name'), attributes['frame_name'],
+                    )
+                break
+            m = QUERY_FRAME_PATTERNS[QUERY_FRAME_PART_DEFAULT].match(line)
+            if m:
+                attributes[DEFAULT_KEY] = m.group(DEFAULT_KEY)
+                continue
+            m = QUERY_FRAME_PATTERNS[QUERY_FRAME_PART_HIDDEN].match(line)
+            if m:
+                if m.group(HIDDEN_KEY) == "TRUE":
+                    attributes[HIDDEN_KEY] = True
+                elif m.group(HIDDEN_KEY) == "FALSE":
+                    attributes[HIDDEN_KEY] = False
+                else:
+                    self.logger.warning(
+                        "The QHidden(%s) has invalid option: "
+                        "%s not in (TRUE, FALSE)", attributes['frame_name'],
+                        m.group(HIDDEN_KEY)
+                    )
+                    attributes[HIDDEN_KEY] = False
+                continue
+            m = QUERY_FRAME_PATTERNS[
+                QUERY_FRAME_PART_VALID_VALUES
+            ].match(line)
+            if m:
+                attributes[VALID_VALUES_KEY] = [
+                    v.replace("\\\\", "\\").replace("\\|", "|")
+                    for v in re.split(r'(?<!\\)\|', m.group("valid"))
```

Comment:
I think the correct way to parse this should be something like:

Create an array of records, each has {char, type} where type is one of {plain, unparsed }

Initialize it with the input string and type=unparsed for all chars.

Then loop over:
Replace each [('\', unparsed), ('\', unparsed)] with ('\', plain)
Replace each [('\', unparsed), ('|', unparsed)] with ('|', plain)
split the result on [('\', unparsed), ('|', unparsed)]

Or something like that, didn't actually try this :-)

It might be possible to do using plain re/string replace, not sure.
It seems to me that if you do not keep per char whether it was already parsed ("plain"), you risk breaking on some corner cases.